router/proxy/reverseproxy.go

```diff
@@ -84,16 +87,47 @@ func (p *ReverseProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	res, err := transport.RoundTrip(outreq)
-	if err != nil {
-		p.logf("router: proxy error: %v", err)
-		rw.WriteHeader(http.StatusServiceUnavailable)
-		rw.Write(serviceUnavailable)
+	ctx, cancel := context.WithCancel(ctx)
+	resultc := make(chan httpResult, 1)
+	clientGone := rw.(http.CloseNotifier).CloseNotify()
+
+	go func() {
+		res, err := transport.RoundTrip(ctx, outreq)
+		defer close(resultc)
+		resultc <- httpResult{res, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		go throwawayResponse(resultc)
+		return
+	case <-clientGone:
+		cancel() // client went away, cancel request
+		go throwawayResponse(resultc)
+		// TODO(bgentry): anything special to log here?
 		return
+	case result := <-resultc:
+		if result.err != nil {
+			if result.err == errRequestCanceled {
+				// TODO(bgentry): anything special to log here?
```

Comment:
Is there anything I should log for these scenarios, or should I just drop the TODOs?